modules/serviceregistry/src/main/java/org/opencastproject/serviceregistry/impl/JobDispatcher.java

```diff
@@ -411,12 +416,29 @@ private void dispatchDispatchableJobs(EntityManager em, List<JpaJob> jobsToDispa
             candidateServices = serviceRegistry.getServiceRegistrationsByLoad(jobType, services, hosts, systemLoad);
           }
 
+          //if enabled use the hardware loads to exclude fully occupied services
+          if (serviceRegistry.isHardwareLoadEnabled()) {
+            final List<ServiceRegistration> servicesToRemove = new ArrayList<ServiceRegistration>();
+
+            for (ServiceRegistration serviceReg : candidateServices) {
+              if (systemLoad.get(serviceReg.getHost()).getCurrentLoad() > systemLoad.get(serviceReg.getHost()).getMaxLoad())
+                servicesToRemove.add(serviceReg);
+            }
+            candidateServices.removeAll(servicesToRemove);
+            if (candidateServices.isEmpty())
+              logger.info("Dispatching of job {} delayed, all potential services of type '{}' are occupied", job, jobType);
+          }
+
           // Try to dispatch the job
           String hostAcceptingJob = null;
           try {
             hostAcceptingJob = dispatchJob(em, job, candidateServices);
             try {
-              systemLoad.updateNodeLoad(hostAcceptingJob, job.getJobLoad());
+              //if hardware load is enabled update the hardware load value for the host
+              if (!serviceRegistry.isHardwareLoadEnabled())
+                systemLoad.updateNodeLoad(hostAcceptingJob, job.getJobLoad());
+              else
+                systemLoad.get(hostAcceptingJob).setCurrentLoad(serviceRegistry.getHardwareLoadbyHost(hostAcceptingJob));
```

Comment:
This is (IMO) quite risky.  You're making the assumption that the dispatched job starts imposing its load immediately, which is not true.  You're also now making a (slow!) HTTP request as part of the dispatch loop for *every single dispatched job*.  I suspect this will have drastic perf implications when deployed at scale.